### DISCUSSION TITLE
EntityRepository.create() persisting entities

### DISCUSSION THREAD

**Discussion by vladkrasn** [mikro-orm/mikro-orm#5897] [General]
It is my understanding that `EntityRepository.create()` automatically persists an entity. So then, you only need to flush the `EntityManager`, and it all works fine.

This is the pattern showcased here https://github.com/mikro-orm/nestjs-example-app/blob/master/src/modules/book/book.controller.ts 

And it's stated that it works that way here https://github.com/mikro-orm/mikro-orm/discussions/4944

It also might be what is being referenced in the second paragraph of https://mikro-orm.io/docs/transactions

However, I can't find this information explicitly stated anywhere in the docs (looked in https://mikro-orm.io/docs/repositories  https://mikro-orm.io/docs/transactions ), and there's nothing about it in the method's JSDoc as well (https://mikro-orm.io/api/core/class/EntityRepository#create).

Would be nice of this gets cleared up, because the whole way MikroORM handles transactions is kinda hard to understand (at least for me), and the fact that there's multiple ways to do the same thing that all *seem* to work fine, make it tough to actually understand how to write code properly.

**Comment by B4nan** [comment]
This is documented in a few places

https://mikro-orm.io/docs/guide/relationships#creating-entity-graph
https://mikro-orm.io/docs/guide/advanced#creating-entities

But you are right that it should be mentioned in the other places too, this is rather "a new behavior", introduced in v5.5 iirc.

I wonder what part you think is hard to understand about the transactions, at the core it's pretty simple, a flush is a transaction.

**Reply by B4nan** [reply]
I've updated the api doc via 03295ae9bc0f6742b8e08401c9f9ed69b2e02887. I don't think this is something we should mention in the repositories section, we already say the important part in there right at the top - a repository is just a thin layer on top of EM, it does not have any functionality really, it's all done by the EM. Similarly, I am not sure why you were searching for this in the transactions section, since there is not a single mention of the `em.create` method.

**Reply by vladkrasn** [reply]
Thank you! I didn't know where to look, so I looked everywhere I could think of.

> I wonder what part you think is hard to understand about the transactions, at the core it's pretty simple, a flush is a transaction.

I guess it's just I'm fuzzy on how certain situations work.

For example, if two requests come in, and they persist their entities at roughly the same time (create them), but then the first request decides to flush the changes, and the other decides against it. Does the first request's flush flushes the persisted entity of the second request? Or does the second request has its own context, and thus the entity persisted in it is lost forever (as it should be) because it wasn't flushed?

My guess it's the latter, but I couldn't find stuff that supports it https://mikro-orm.io/docs/entity-manager

**Reply by B4nan** [reply]
Two requests should have two contexts, they are distinct. This is usually handled through a request context middleware.

https://mikro-orm.io/docs/identity-map#request-context
https://mikro-orm.io/docs/guide/project-setup#fastify

If both will flush, its the database that deal with the actual outcome, and it depends on several things, like the transaction isolation level (which you can set with explicit transaction demarcation, see https://mikro-orm.io/docs/transactions#isolation-levels). Or you could use optimistic locking, see https://mikro-orm.io/docs/transactions#optimistic-locking.
